File: resources/baidu_sign.js
function toArray(text) {
    if (Array.isArray(text)) {
        let result = Array(text.length);
        for (let i = 0; i < text.length; i++) {
            result[i] = text[i];
        }

        return result
    }

    return Array.from(text)
}

function encodeByKey(num, key) {
    for (let t = 0; t < key.length - 2; t += 3) {
        let a = key[t + 2];

        if (a >= 'a') {
            a = a.charCodeAt(0) - 87;
        } else {
            a = Number(a);
        }


        if (key[t + 1] === '+') {
            a = num >>> a;
        } else {
            a = num << a;
        }


        if ("+" === key[t]) {
            num = num + a & 4294967295;
        } else {
            num = num ^ a;
        }
    }

    return num
}

function generateToken(text, gtk) {
    let outUtf8 = text.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]/g);
    if (null === outUtf8) {
        if (text.length > 30) {
            text =
                text.substr(0, 10) +
                text.substr(Math.floor(text.length / 2) - 5, 10) +
                text.substr(-10, 10)
        }
    } else {
        let chars = text.split(/[\uD800-\uDBFF][\uDC00-\uDFFF]/);
        let charsLength = chars.length;
        let reforms = [];

        for (let i = 0; charsLength > i; i++) {
            if ("" !== chars[i]) {
                reforms.push.apply(reforms, toArray(chars[i].split("")))
            }

            if (i !== charsLength - 1) {
                reforms.push(outUtf8[i]);
            }
        }

        if (reforms.length > 30) {
            text =
                reforms.slice(0, 10).join("") +
                reforms.slice(
                    Math.floor(reforms.length / 2) - 5,
                    Math.floor(reforms.length / 2) + 5
                ).join("") +
                reforms.slice(-10).join("")
        }
    }
    let gtkArray = gtk.split(".");
    let prefix = Number(gtkArray[0]) || 0;
    let suffix = Number(gtkArray[1]) || 0;
    let s = [];

    for (let v = 0; v < text.length; v++) {
        let a = text.charCodeAt(v);
        if (128 > a) {
            s.push(a)
        } else {
            if (2048 > a) {
                s.push(a >> 6 | 192);
            } else {
                if (55296 === (64512 & a) && v + 1 < text.length && 56320 === (64512 & text.charCodeAt(v + 1))) {
                    a = 65536 + ((1023 & a) << 10) + (1023 & text.charCodeAt(++v));
                    s.push(a >> 18 | 240);
                    s.push(a >> 12 & 63 | 128);
                } else {
                    s.push(a >> 12 | 224)
                }

                s.push(a >> 6 & 63 | 128)
            }

            s.push(63 & a | 128)
        }
    }

    let key1 = "+-a^+6";
    let key2 = "+-3^+b+-f";
    let p = prefix;
    for (let b = 0; b < s.length; b++) {
        p += s[b];
        p = encodeByKey(p, key1);
    }

    p = encodeByKey(p, key2);
    p ^= suffix;

    if (p < 0) {
        p = (2147483647 & p) + 2147483648
    }

    p %= 1e6;

    return p.toString() + "." + (p ^ prefix)
}

// console.log(encodeByKey(123456, '+-a^6'));
const gtk = '320305.131321201';
console.log('😀', generateToken('😀', gtk) === "137522.457219");
console.log(generateToken('hello', gtk) === '54706.276099');
console.log(generateToken('hello world are you ok ? this is funny.', gtk));
console.log(generateToken('你好', gtk));
console.log(generateToken('抽刀断水水更流，举杯消愁愁更愁。', gtk));
// console.log(encodeByKey(320305, '+-a^+6'));

File: translate/util/baidu_sign.py
import math

__all__ = ['generate_token']


def encode_by_key(num, key):
    t = 0
    while t < len(key) - 2:
        a = key[t + 2]
        if ord(a) >= ord('a'):
            a = ord(a) - 87
        else:
            a = int(a)

        if key[t + 1] == '+':
            a = num >> a
        else:
            a = num << a

        if '+' == key[t]:
            num = num + a & 4294967295
        else:
            num = num ^ a

        t += 3
    return num


def substr(string, start, length):
    end = start + length
    if start < 0:
        end = length
    return string[start:end]


def generate_token(text, gtk):
    if len(text) > 30:
        text = substr(text, 0, 10) + substr(text, math.floor(len(text) / 2) - 5, 10) + substr(text, -10, len(text))

    gtk_list = gtk.split('.')
    prefix = int(gtk_list[0])
    suffix = int(gtk_list[1])
    s = []
    v = 0

    while v < len(text):
        a = ord(text[v])

        if 128 > a:
            s.append(a)
        else:
            if 2048 > a:
                s.append(a >> 6 | 192)
            else:
                if 55296 == (64512 & a) and v + 1 < text.length and 56320 == (64512 & ord(text[v + 1])):
                    a = 65536 + ((1024 & a) << 10) + (1023 & ord(text[++v]))
                    s.append(a >> 18 | 240)
                    s.append(a >> 12 & 63 | 128)
                else:
                    s.append(a >> 12 | 224)

                s.append(a >> 6 & 63 | 128)

            s.append(63 & a | 128)
        v += 1

    key1 = '+-a^+6'
    key2 = "+-3^+b+-f"
    p = prefix

    b = 0
    while b < len(s):
        p += s[b]
        p = encode_by_key(p, key1)
        b += 1

    p = encode_by_key(p, key2)
    p ^= suffix

    if p < 0:
        p = (2147483647 & p) + 2147483648

    p %= 1e6
    p = int(p)

    return str(p) + '.' + str(p ^ prefix)
